devicetypes/smartthings/zigbee-battery-accessory-dimmer.src/zigbee-battery-accessory-dimmer.groovy

```diff
@@ -47,38 +48,83 @@ metadata {
 	}
 }
 
-def getDOUBLE_STEP() {10}
-def getSTEP() {5}
+def getDOUBLE_STEP() { 10 }
+def getSTEP() { 5 }
 
-def parse(String description) {
+def getONOFF_ON_COMMAND() { 0x0001 }
+def getONOFF_OFF_COMMAND() { 0x0000 }
+def getLEVEL_MOVE_LEVEL_COMMAND() { 0x0000 }
+def getLEVEL_MOVE_COMMAND() { 0x0001 }
+def getLEVEL_STEP_COMMAND() { 0x0002 }
+def getLEVEL_STOP_COMMAND() { 0x0003 }
+def getLEVEL_MOVE_LEVEL_ONOFF_COMMAND() { 0x0004 }
+def getLEVEL_MOVE_ONOFF_COMMAND() { 0x0005 }
+def getLEVEL_STEP_ONOFF_COMMAND() { 0x0006 }
+def getLEVEL_STOP_ONOFF_COMMAND() { 0x0007 }
+def getLEVEL_DIRECTION_UP() { "00" }
+def getLEVEL_DIRECTION_DOWN() { "01" }
 
-	def descMap = zigbee.parseDescriptionAsMap(description)
+def getBATTERY_VOLTAGE_ATTR() { 0x0020 }
+def getBATTERY_PERCENT_ATTR() { 0x0021 }
 
-	if (isSengledSwitch()) {
-		handleSengledSwitchEvents(descMap)
-	} else if (isCentraliteSwitch())  {
-		handleCentraliteSmartSwitchEvents(descMap)
-	}
-	handleBatteryEvents(descMap)
-}
+def getMFR_SPECIFIC_CLUSTER() { 0xFC10 }
+
+def getUINT8_STR() { "20" }
 
+
+private boolean isIkeaDimmer() {
+	device.getDataValue("model") == "TRADFRI wireless dimmer"
+}
 private boolean isSengledSwitch() {
 	device.getDataValue("model") == "E1E-G7F"
 }
-
 private boolean isCentraliteSwitch() {
 	device.getDataValue("model") == "3131-G"
 }
 
+def parse(String description) {
+	log.debug "description is $description"
+	def results = []
+
+	def event = zigbee.getEvent(description)
+	if (event) {
+		results << createEvent(event)
+	} else {
+		def descMap = zigbee.parseDescriptionAsMap(description)
+
+		if (isSengledSwitch()) {
+			results = handleSengledSwitchEvents(descMap)
+		} else if (descMap.clusterInt == zigbee.ONOFF_CLUSTER) {
+			results = handleSwitchEvent(descMap)
+		} else if (descMap.clusterInt == zigbee.LEVEL_CONTROL_CLUSTER) {
+			if (isCentraliteSwitch()) {
+				results = handleCentraliteSmartSwitchLevelEvent(descMap)
+			} else if (isIkeaDimmer()) {
+				results = handleIkeaDimmerLevelEvent(descMap)
+			}
+		} else if (descMap.clusterInt == zigbee.POWER_CONFIGURATION_CLUSTER) {
+			results = handleBatteryEvents(descMap)
+		} else {
+			log.warn "DID NOT PARSE MESSAGE for description : $description"
+			log.debug "${descMap}"
+		}
+	}
+
+	log.debug "parse returned $results"
+	return results
+}
+
 def handleSengledSwitchEvents(descMap) {
-	if (descMap?.clusterInt == 0xFC10) {
+	def results = []
+
+	if (descMap?.clusterInt == MFR_SPECIFIC_CLUSTER && descMap.data) {
 		def currentLevel = device.currentValue("level") as Integer ?: 0
-		def value = 0
+		def value = currentLevel
```

Comment:
I'm thinking that I should leave this and add a debug print in the '02' and '03' cases below if descMap.data[2] is not what we are expecting.

Or

Just have a default action, like
```
if (descMap.data[2] == '02') {

} else { // descMap.data[2] == '01' or any other unexpected value

}
```

Any thoughts?

Reply:
Yeah I'd agree with leave it as is here and handle the unexpected value with a log message and setting `level` to `currentLevel` at that point

Reply:
Done!